### PR TITLE
Migrate to ChannelBuffers.push

### PR DESCRIPTION
**Pull request by Hixie**
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Hixie** [comment]
Blocked on https://github.com/dart-lang/dartdoc/issues/2526
Blocken on https://github.com/flutter/flutter/pull/76344

**Comment by Hixie** [comment]
Blocked on https://github.com/flutter/website/pull/5322

**Comment by Hixie** [comment]
Blocked on plugins referring to the old APIs (just need to import flutter_test).

**Comment by Hixie** [comment]
Blocked on having the deprecation warnings commented out.

**Comment by Hixie** [comment]
cc @goderbauer the tests finally pass, would love your review!

**Comment by Hixie** [comment]
When reviewing this, looking at the commits separately should make review significantly easier.

**Review by goderbauer — COMMENTED** [review]
I do like that all that mocking stuff is no longer in the framework.

File: packages/flutter/test/material/feedback_tester.dart
Comment (goderbauer) [reply packages/flutter/test/material/feedback_tester.dart]: Ignore this, I was reviewing the PR one commit at a time and just now realize that this if fixed in a later commit.

Comment (Hixie) [reply packages/flutter/test/material/feedback_tester.dart]: the `deprecated.dart` extension makes this work (and then i avoid using it in the later commit)

File: packages/flutter/test/rendering/editable_test.dart
Comment (Hixie) [reply packages/flutter/test/rendering/editable_test.dart:32]: there should, but changing that is part of my much bigger `binding` fix in a later PR. I can remove this line if you like, I just needed it to make sure I wasn't creating the wrong binding when debugging the tests.

File: .cirrus.yml
Comment (Hixie) [reply .cirrus.yml:194]: oops, that slipped in from another PR i was working on. will remove.

File: packages/flutter_test/lib/src/binding.dart
Comment (Hixie) [reply packages/flutter_test/lib/src/binding.dart:110]: it would have made the diff harder to review. that was before i realised i was going to have to have multiple commits anyway, though, so i could always add it here in its own commmit.

File: packages/flutter/lib/src/services/binary_messenger.dart
Comment (Hixie) [reply packages/flutter/lib/src/services/binary_messenger.dart:19]: added more documentation

File: packages/flutter/lib/src/services/binding.dart
Comment (Hixie) [reply packages/flutter/lib/src/services/binding.dart:50]: no, the other direction is done via `PlatformDispatcher.sendPlatformMessage`. Added a "see also".

File: packages/flutter/lib/src/services/binding.dart
Comment (Hixie) [reply packages/flutter/lib/src/services/binding.dart:308]: weird, why did the analyzer not catch that?
removed the `if`.

File: packages/flutter/lib/src/services/binding.dart
Comment (Hixie) [reply packages/flutter/lib/src/services/binding.dart:255]: Where is it called `message`?

Comment (Hixie) [reply packages/flutter/lib/src/services/binding.dart:255]: I found places where ByteData objects were called message and others where they were called data and we could standardise on one, but I'd really rather not do that in this PR given the scope of the PR already. It seems like all the handlePlatformMessage implementations use "data".

File: packages/flutter/test/material/feedback_tester.dart
Comment (Hixie) [reply packages/flutter/test/material/feedback_tester.dart:25]: done

File: packages/flutter/lib/src/services/binding.dart
Comment (Hixie) [reply packages/flutter/lib/src/services/binding.dart:60]: added a parenthetical

File: packages/flutter/test/services/raw_keyboard_test.dart
Comment (Hixie) [reply packages/flutter/test/services/raw_keyboard_test.dart:634]: fixed

File: packages/flutter_test/lib/src/binding.dart
Comment (Hixie) [reply packages/flutter_test/lib/src/binding.dart:146]: Added a TODO to remove the override and leave the method.

I'd love for us to rename this method (and simplify the API) but let's do that in a different PR (maybe add it to #24722). This PR is already hard enough to think about as it is...

Comment (Hixie) [reply packages/flutter_test/lib/src/binding.dart:146]: (i'd love to actually remove this method entirely, and originally i had deprecated it, but too many tests would need to be significantly redesigned. we really depend on being able to wait for the message to have been handled.)

Comment (Hixie) [reply packages/flutter_test/lib/src/binding.dart:146]: done

File: packages/flutter/lib/src/services/binary_messenger.dart
Comment (Hixie) [reply packages/flutter/lib/src/services/binary_messenger.dart:32]: we have to migrate the customer tests, but they can't be migrated until the new API lands

Comment (Hixie) [reply packages/flutter/lib/src/services/binary_messenger.dart:32]: (specifically cocoon)

File: packages/flutter_test/lib/src/deprecated.dart
Comment (Hixie) [reply packages/flutter_test/lib/src/deprecated.dart:23]: as above, once cocoon is migrated to the API added by this PR

File: packages/flutter/lib/src/services/binary_messenger.dart
Comment (goderbauer) [reply packages/flutter/lib/src/services/binary_messenger.dart:45]: Maybe add this to the comment to share that knowledge with people who run across this code before the migration has happened, e.g. "depreciate this method after customers (specifically cocoon) have been migrated:"

File: packages/flutter/test/rendering/editable_test.dart
Comment (goderbauer) [reply packages/flutter/test/rendering/editable_test.dart:32]: Maybe let's just remove it? (or turn it into an assert: `assert(renderer is TestRenderingFlutterBinding)`)

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/services/binding.dart
Comment (goderbauer) [reply packages/flutter/lib/src/services/binding.dart]: Probably because we disable that check to allow for `assert(foo != null);` code even when foo is non-nullable:

https://github.com/flutter/flutter/blob/efbde443d2a6f810d0ceb6097fc79ed0b30bddce/analysis_options.yaml#L39-L40

Comment (Hixie) [reply packages/flutter/lib/src/services/binding.dart]: aaah, d'oh.

File: packages/flutter/test/rendering/editable_test.dart
Comment (Hixie) [reply packages/flutter/test/rendering/editable_test.dart:32]: will remove

File: packages/flutter/lib/src/services/binary_messenger.dart
Comment (Hixie) [reply packages/flutter/lib/src/services/binary_messenger.dart:45]: done

**Comment by Hixie** [comment]
Applied changes.

Blocked on making sure it passes Google's existing tests (I'll have to add flutter_test imports in various places probably).

**Comment by Hixie** [comment]
I merged this to top of tree. Waiting for https://github.com/flutter/flutter/pull/79061 to roll into google3 so that I can test this there more thoroughly.

**Comment by Hixie** [comment]
Remerged to top of tree, waiting for #80003 to land in google3 for further testing.

**Comment by Hixie** [comment]
I'm landing this now since everything is green, but I fully expect this to get reverted...

**Comment by Hixie** [comment]
Interestingly some systems at Google reckon that this reduced codesize by 1-2%, which is an unexpected but nice side effect.

**Pull request merged**
